### PR TITLE
feat(rules): support ranges as rule conditions

### PR DESCRIPTION
**Pull request by boc-tothefuture**
This PR supports ranges as rule conditions for from and to

Additionally, to support this some refactoring was done:
1. Change from @trigger_delays to @trigger_conditions 
2. Simplification of automation rule, but making TriggerDelay a TriggerCondition and moving it into its own class

### CODE REVIEW

File: docs/usage/triggers/changed.md [hunk 1]
```diff
@@ -20,7 +20,9 @@ grand_parent: Usage
 Changed accepts Items, Things or Groups. 
 To and from accept arrays to match multiple states
 
-The from and to values operate exactly as they do in the DSL and Python rules with the exception of operating on Things.  If changed element being used as a trigger is a thing than the to and from values will accept symbols and strings, where the symbol matches the [supported status](https://www.openhab.org/docs/concepts/things.html). 
+The from and to values operate like they do in the DSL and Python rules with some enhabnacements.
+1. If changed element being used as a trigger is a thing than the to and from values will accept symbols and strings, where the symbol matches the [supported status](https://www.openhab.org/docs/concepts/things.html). 
+2. Support for ranges
```
Comment (jimtng) [inline docs/usage/triggers/changed.md:25]:
```suggestion
The from and to values operate like they do in the DSL and Python rules with some enhancements:
1. If the changed element being used as a trigger is a thing, the `to` and `from` values will accept symbols and strings, where the symbol matches the [supported status](https://www.openhab.org/docs/concepts/things.html). 
2. Support for ranges.
```

File: docs/usage/triggers/updated.md [hunk 2]
```diff
@@ -17,7 +17,7 @@ grand_parent: Usage
 
 Changed accepts Items, Things or Groups. 
 
-The to value restricts the rule from running to only if the updated state matches. If the updated element being used as a trigger is a thing than the to and from values will accept symbols and strings, where the symbol matches the [supported status](https://www.openhab.org/docs/concepts/things.html). 
+The `to` value restricts the rule from running to only if the updated state matches. If the updated element being used as a trigger is a thing than the to and from values will accept symbols and strings, where the symbol matches the [supported status](https://www.openhab.org/docs/concepts/things.html).  The `to` value supports ranges.
```
Comment (jimtng) [inline docs/usage/triggers/updated.md:20]:
```suggestion
The `to` value restricts the rule from running only if the updated state matches. If the updated element being used as a trigger is a thing, the `to` and `from` values will accept symbols and strings, where the symbol matches the [supported status](https://www.openhab.org/docs/concepts/things.html).  The `to` value supports ranges.
```

File: lib/openhab/dsl/rules/triggers/conditions/range.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            # Process rule
+            # @param [Hash] inputs inputs from trigger
+            #
+            def process(mod:, inputs:) # rubocop:disable Lint/UnusedMethodArgument - mod is unused here but required
+              logger.trace("Checking #{inputs} against condition trigger #{self}")
+              yield if check_from(inputs: inputs) && check_to(inputs: inputs)
+            end
+
+            #
+            # Check if from condition match the in inputs
```
Comment (jimtng) [inline lib/openhab/dsl/rules/triggers/conditions/range.rb:36]:
```suggestion
            # Check if from condition match the inputs
```

**Review by jimtng — CHANGES_REQUESTED** [review]
Great work! You must have read my mind. I was hesitant to raise the request for this feature, but I'm glad you went ahead and implemented it. This will simplify some of my rules.

I'd suggest adding some more tests that involve the edges of the range and also ranges with excluded end.

**Comment by boc-tothefuture** [comment]
> Great work! You must have read my mind. I was hesitant to raise the request for this feature, but I'm glad you went ahead and implemented it. This will simplify some of my rules.
> 
> I'd suggest adding some more tests that involve the edges of the range and also ranges with excluded end.

I did it after looking at the fridge temperature rule you posted -- although this won't solve that specific case. 

Did you have any specific tests in mind? I wasn't trying to test if the Ruby Range class works, just that we process ranges correctly.

**Comment by jimtng** [comment]
> I did it after looking at the fridge temperature rule you posted -- although this won't solve that specific case.

Exactly, that's the rule that made me want to have this feature. I was just thinking about how to use this for that rule, perhaps in combination with `otherwise`. I was also thinking of using `for` (changed duration feature) but that would require the inverse of this PR, i.e. 
```rb
changed Fridge_Temperature, not_to: '2 °C'..'6 °C', for: 5.minutes
```

This feature will also be handy to check for SunElevation. I have a rule that gradually changes a light's colour temperature  based on changes in SunElevation during the twilight hours.



> Did you have any specific tests in mind? I wasn't trying to test if the Ruby Range class works, just that we process ranges correctly.

Nothing specific, and I understand that this is not specifically for testing the range itself. Happy to drop this request.

**Comment by jimtng** [comment]
I currently implemented my colour temperature changing rule like this:
```rb
rule 'Dawn/Dusk: Change light temperature' do
  changed Sun_Elevation
  only_if { ELEVATION_RANGE.cover? Sun_Elevation }
```

**Comment by boc-tothefuture** [comment]
> > I did it after looking at the fridge temperature rule you posted -- although this won't solve that specific case.
> 
> Exactly, that's the rule that made me want to have this feature. I was just thinking about how to use this for that rule, perhaps in combination with `otherwise`. I was also thinking of using `for` (changed duration feature) but that would require the inverse of this PR, i.e.
> 
> ```ruby
> changed Fridge_Temperature, not_to: '2 °C'..'6 °C', for: 5.minutes
> ```
> 
> This feature will also be handy to check for SunElevation. I have a rule that gradually changes a light's colour temperature based on changes in SunElevation during the twilight hours.
> 
> > Did you have any specific tests in mind? I wasn't trying to test if the Ruby Range class works, just that we process ranges correctly.
> 
> Nothing specific, and I understand that this is not specifically for testing the range itself. Happy to drop this request.

Let's move this part to another issue.  I will merge the range here and we can talk about other implementations in another issue.

**Pull request merged**